lib/timeout.rb

```diff
@@ -50,9 +50,74 @@ def exception(*)
   end
 
   # :stopdoc:
-  THIS_FILE = /\A#{Regexp.quote(__FILE__)}:/o
-  CALLER_OFFSET = ((c = caller[0]) && THIS_FILE =~ c) ? 1 : 0
-  private_constant :THIS_FILE, :CALLER_OFFSET
+  CONDVAR = ConditionVariable.new
+  QUEUE = Queue.new
+  TIMEOUT_THREAD_MUTEX = Mutex.new
+  @timeout_thread = nil
+  private_constant :CONDVAR, :QUEUE, :TIMEOUT_THREAD_MUTEX
+
+  class Request
+    attr_reader :deadline
+
+    def initialize(thread, timeout, exception_class, message)
+      @thread = thread
+      @deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
+      @exception_class = exception_class
+      @message = message
+
+      @mutex = Mutex.new
+    end
+
+    def done?
+      @mutex.locked?
+    end
+
+    def expired?(now)
+      now >= @deadline and !done?
+    end
+
+    def interrupt
+      if @mutex.try_lock
+        @thread.raise @exception_class, @message
+      end
+    end
+
+    def finished
+      @mutex.try_lock
+    end
+  end
+  private_constant :Request
+
+  def self.ensure_timer_thread_created
+    unless @timeout_thread
+      TIMEOUT_THREAD_MUTEX.synchronize do
+        @timeout_thread ||= Thread.new do
+          mutex = Mutex.new
+          requests = []
+          while true
+            until QUEUE.empty? and !requests.empty? # wait to have at least one request
+              req = QUEUE.pop
+              requests << req
+            end
+            closest_deadline = requests.min_by(&:deadline).deadline
+
+            now = 0.0
+            mutex.synchronize do
```

Comment:
We only do this because ConditionVariable#wait needs to be given a Mutex.
We need no synchronization from the ConditionVariable/Mutex though, the Queue and Request are both thread-safe on their own.

Reply:
Actually, we need to prevent the CV#signal to be done before the CV#wait otherwise it will be missed.
So we use the same Mutex here and on push to ensure the signal is always seen, or if it's sent before, then it's before this section and the QUEUE.empty? check.